File: js_edition/src/tests/absences.spec.js
import chai from 'chai';
import chaiHttp from 'chai-http';
import app from '../app';

chai.use(chaiHttp);
chai.should();

const testValidAbsencesRoute = (route, test) => {
  return chai.request(app).get(route).end((err, res) => {
    res.should.have.status(200);
    res.body.should.have.property('total');
    res.body.should.have.property('absences');
    test(err, res);
  });
};

const testInvalidAbsencesRoute = (route, test) => {
  return chai.request(app).get(route).end((err, res) => {
    res.body.should.have.property('error');
    test(err, res);
  });
};

describe('Absences', () => {
  describe('GET /absences', () => {
    it('should get all absences records with employees names', (done) => {
      const test = (err, res) => {
        res.body.absences
          .filter((e) => e.hasOwnProperty('userName'))
          .should.have.lengthOf(res.body.total);
        done();
      };
      testValidAbsencesRoute('/absences', test);
    });

    it('should get absences from a valid user id', (done) => {
      const id = 2664;
      const test = (err, res) => {
        res.body.absences
          .filter((e) => e.userId === id)
          .should.have.lengthOf(res.body.total);
        done();
      };
      testValidAbsencesRoute(`/absences?userId=${id}`, test);
    });

    it('should not get absences from a valid user id that has no absence', (done) => {
      const id = 2290;
      const test = (err, res) => {
        res.should.have.status(404);
        done();
      };
      testInvalidAbsencesRoute(`/absences?userId=${id}`, test);
    });

    it('should not get absences from a invalid user id', (done) => {
      const id = -1;
      const test = (err, res) => {
        res.should.have.status(404);
        done();
      };
      testInvalidAbsencesRoute(`/absences?userId=${id}`, test);
    });

    it('should get absences from a given date range', (done) => {
      const [startDate, endDate] = ['2017-01-01', '2017-02-01'];
      const test = (err, res) => {
        const filterStart = new Date().setUTCFullYear('2017', '01', '01');
        const filterEnd = new Date().setUTCFullYear('2017', '02', '01');
        res.body.absences
          .filter((e) => {
            const [startYear, startMonth, startDay] = e.startDate.split('-');
            const [endYear, endMonth, endDay] = e.endDate.split('-');
            const absStart = new Date().setUTCFullYear(startYear, startMonth, startDay);
            const absEnd = new Date().setUTCFullYear(endYear, endMonth, endDay);
            return (absStart >= filterStart && absStart <= filterEnd)
              || (absEnd >= filterStart && absEnd <= filterEnd);
          })
          .should.have.lengthOf(res.body.total);
        done();
      };
      testValidAbsencesRoute(`/absences?startDate=${startDate}&endDate=${endDate}`, test);
    });

    it('should not get absences from a invalid date range', (done) => {
      const [startDate, endDate] = ['2017-01-01', '2016-08-01'];
      const test = (err, res) => {
        res.should.have.status(404);
        done();
      };
      testInvalidAbsencesRoute(`/absences?startDate=${startDate}&endDate=${endDate}`, test);
    });

    it('should get absences from a given date range and valid user id', (done) => {
      const [startDate, endDate] = ['2017-01-01', '2017-02-01'];
      const id = 2664;
      const test = (err, res) => {
        const filterStart = new Date().setUTCFullYear('2017', '01', '01');
        const filterEnd = new Date().setUTCFullYear('2017', '02', '01');
        res.body.absences
          .filter((e) => {
            const [startYear, startMonth, startDay] = e.startDate.split('-');
            const [endYear, endMonth, endDay] = e.endDate.split('-');
            const absStart = new Date().setUTCFullYear(startYear, startMonth, startDay);
            const absEnd = new Date().setUTCFullYear(endYear, endMonth, endDay);
            return (absStart >= filterStart && absStart <= filterEnd)
              || (absEnd >= filterStart && absEnd <= filterEnd);
          })
          .filter((e) => e.userId === id)
          .should.have.lengthOf(res.body.total);
        done();
      };
      testValidAbsencesRoute(`/absences?startDate=${startDate}&endDate=${endDate}&userId=${id}`, test);
    });
  });

  describe('GET /absences/sickness', () => {
    it('should get absences due to sickness', (done) => {
      const test = (err, res) => {
        res.body.absences
          .filter((e) => e.message.includes('sick'))
          .should.have.lengthOf(res.body.total);
        done();
      };
      testValidAbsencesRoute('/absences/sickness', test);
    });
  });

  describe('GET /absences/vacations', () => {
    it('should get absences due to vacation', (done) => {
      const test = (err, res) => {
        res.body.absences
          .filter((e) => e.message.includes('vacation'))
          .should.have.lengthOf(res.body.total);
        done();
      };
      testValidAbsencesRoute('/absences/vacations', test);
    });
  });
});

describe('Calendar', () => {
  describe('GET /', () => {
    it('should download the absences in .ics format', (done) => {
      chai.request(app)
        .get('/')
        .end((err, res) => {
          res.should.have.status(200);
          done();
        });
    });
  });
});

File: ruby_edition/spec/absences_spec.rb
# frozen_string_literal: true

require_relative '../server'
require 'rack/test'
require 'date'

RSpec.describe 'The Absence Manager' do
  include Rack::Test::Methods

  def app
    Sinatra::Application
  end

  it 'should get all absences records with employees names' do
    get '/absences'
    expect(last_response).to be_ok
    response = JSON.parse(last_response.body)
    response['absences'] = response['absences'].select {|abs| abs.key?('userName')}
    expect(response['absences'].length).to eq(response['total'])
  end

  it 'should get absences from a valid user id' do
    get '/absences?userId=2664'
    response = JSON.parse(last_response.body)
    response['absences'] = response['absences'].select {|abs| abs['userId'] == 2664}
    expect(response['absences'].length).to eq(response['total'])
  end

  it 'should not get absences from a valid user id that has no absence' do
    get '/absences?userId=2290'
    expect(last_response).to be_ok
    response = JSON.parse(last_response.body)
    expect(response["absences"].length).to eq(0)
  end

  it 'should not get absences from a invalid user id' do
    get '/absences?userId=-1'
    expect(last_response).to be_ok
    response = JSON.parse(last_response.body)
    expect(response["absences"].length).to eq(0)
  end

  it 'should get absences from a given date range' do
    get '/absences?startDate=2017-01-01&endDate=2017-02-01'
    expect(last_response).to be_ok
    response = JSON.parse(last_response.body)
    filter_start = Date.new(2017, 01, 01)
    filter_start = Date.new(2017, 02, 01)
    response['absences'] = response['absences'] do select { |abs|
      date = abs['startDate'].split('-')
      start_date = Date.new(date[0].to_i, date[1].to_i, date[2].to_i)
      date = abs['endDate'].split('-')
      end_date = Date.new(date[0].to_i, date[1].to_i, date[2].to_i)
      ((start_date >= filter_start && start_date <= filter_end) ||
        (end_date >= filter_start && end_date <= filter_end))
    }
    end
    expect(response['absences'].length).to eq(response['total'])
  end

  it 'should not get absences from a invalid date range' do
    get '/absences?startDate=2017-01-01&endDate=2016-08-01'
    expect(last_response).to be_ok
    response = JSON.parse(last_response.body)
    expect(response["absences"]).to eq([])
  end

  it 'should get absences from a given date range and valid user id' do
    get '/absences?startDate=2017-01-01&endDate=2017-02-01&userId=2664'
    expect(last_response).to be_ok
    response = JSON.parse(last_response.body)
    filter_start = Date.new(2017, 01, 01)
    filter_end = Date.new(2017, 02, 01)
    response['absences'] = response['absences'].select { |abs|
      date = abs['startDate'].split('-')
      start_date = Date.new(date[0].to_i, date[1].to_i, date[2].to_i)
      date = abs['endDate'].split('-')
      end_date = Date.new(date[0].to_i, date[1].to_i, date[2].to_i)
      ((start_date >= filter_start && start_date <= filter_end) ||
        (end_date >= filter_start && end_date <= filter_end))
    }.select {|abs| abs['userId'] == 2664}
    expect(response['absences'].length).to eq(response['total'])
  end

  it 'should get absences due to sickness' do
    get '/absences/sickness'
    expect(last_response).to be_ok
    response = JSON.parse(last_response.body)
    response['absences'].select {|abs| abs['message'].include?('sick')}
    response['absences'].select {|abs| abs['type'] == 'sick'}
    expect(response['absences'].length).to eq(response['total'])
  end

  it 'should get absences due to vacation' do
    get '/absences/vacations'
    expect(last_response).to be_ok
    response = JSON.parse(last_response.body)
    response['absences'].select {|abs| abs['message'].include?('vacation')}
    response['absences'].select {|abs| abs['type'] == 'vacation'}
    expect(response['absences'].length).to eq(response['total'])
  end

  it 'should download the absences in .ics format' do
    get '/'
    expect(last_response).to be_ok
  end


end
